File: androidApp/src/main/java/com/example/rewardcards/android/cardList/CardListViewModel.kt
package com.example.rewardcards.android.cardList

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.rewardcards.domain.card.Card
import com.example.rewardcards.domain.card.CardDataSource
import com.example.rewardcards.domain.card.SearchCards
import com.example.rewardcards.domain.time.DateTimeUtil
import com.example.rewardcards.presentation.BabyBlueHex
import com.example.rewardcards.presentation.RedOrangeHex
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class CardListViewModel @Inject constructor(
    private val cardDataSource: CardDataSource,
    private val savedStateHandle: SavedStateHandle
): ViewModel() {

    private val searchCards = SearchCards()

    private val cards = savedStateHandle.getStateFlow("cards", emptyList<Card>())
    private val searchText = savedStateHandle.getStateFlow("searchText", "")
    private val isSearchActive = savedStateHandle.getStateFlow("isSearchActive", false)

    val state = combine(cards, searchText, isSearchActive) { cards, searchText, isSearchActive ->
        CardListState(
            cards = searchCards.execute(cards, searchText),
            searchCards = searchText,
            isSearchActive = isSearchActive
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), CardListState())

    init {
        viewModelScope.launch {
            if (cardDataSource.getAllCards().isEmpty()) {
                (1..2).forEach {
                    cardDataSource.insertCard(
                        Card(
                            id = null,
                            name = "McDonalds",
                            image = "null",
                            barcode = "69010140244368707",
                            color = RedOrangeHex,
                            created = DateTimeUtil.now(),
                            notes = BabyBlueHex.toString()
                        )
                    )
                }
            }
        }
    }

    fun loadCards() {
        viewModelScope.launch {
            savedStateHandle["cards"] = cardDataSource.getAllCards()
        }
    }

    fun onSearchTextChange(text: String) {
        savedStateHandle["searchText"] = text
    }

    fun onToggleSearch() {
        savedStateHandle["isSearchActive"] = !isSearchActive.value

        if (!isSearchActive.value) {
            savedStateHandle["searchText"] = ""
        }
    }

    fun deleteCardById(id: Long) {
        viewModelScope.launch {
            cardDataSource.deleteCardById(id)
            loadCards()
        }
    }
}

File: iosApp/iosApp/cardList/CardListViewModel.swift
//
//  CardListViewModel.swift
//  iosApp
//
//  Created by Toby on 1/1/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

extension CardListScreen {
    @MainActor class CardListViewModel: ObservableObject {
        private var cardDataSource: CardDataSource? = nil
        
        private let searchCards = SearchCards()
        
        private var cards = [Card]()
        @Published private(set) var filteredCards = [Card]()
        @Published var searchText = "" {
            didSet {
                self.filteredCards = searchCards.execute(cards: self.cards, query: searchText)
            }
        }
        @Published private(set) var isSearchActive = false
        
        init(cardDataSource: CardDataSource? = nil) {
            self.cardDataSource = cardDataSource
        }
        
        func addData() {
            cardDataSource?.insertCard(card: Card(id: nil, name: "Costco", image: "", barcode: "123456789", type: 1, color: 0xFF2341, created: DateTimeUtil().now(), notes: "No notes!"), completionHandler: { error in })
            cardDataSource?.insertCard(card: Card(id: nil, name: "Walmart", image: "", barcode: "4324234421", type: 1, color: 0xFF6262, created: DateTimeUtil().now(), notes: "No notes!"), completionHandler: { error in })
            self.loadCards()
        }
        
        func loadCards() {
            cardDataSource?.getAllCards(completionHandler: { cards, error in
                self.cards = cards ?? []
                self.filteredCards = self.cards
            })
        }
        
        func deleteCardById(id: Int64?) {
            if id != nil {
                cardDataSource?.deleteCardById(id: id!, completionHandler: { error in
                    self.loadCards()
                })
            }
        }
        
        func toggleIsSearchActive() {
            isSearchActive = !isSearchActive
            
            if !isSearchActive {
                searchText = ""
            }
        }
        
        func setCardDataSource(cardDataSource: CardDataSource) {
            self.cardDataSource = cardDataSource
            loadCards()
            //addData()
        }
    }
}
